# code repository

coexpression, differential and cluster analysis used in my masters

### functions

contains all functions and scripts that master scripts direct to produce results

### masterscripts

all scripts that drive the core transcriptional profiling of cell subsets using coexpression and differential expression

| file | function |
| ------ | ----------- |
| coexpression_method_comparison_master.r   | compares three co-expression methods (self organising maps, weighted gene co-expression analysis and k-means) |
| differential_expression_master.r | three different differential expression methodologies |
| kallisto.sh    | raw sequence read alignment and mapping using fasta paired end sequencefiles  |
| self_organising_map_.r    | self-organising map to transcriptionally profile cell subsets  |
| wgcna_master.r    | self-organising map to transcriptionally profile cell subsets  |




