# code repository

coexpression, differential and cluster analysis used in my masters

## contents
'masters_scripts_for_masters'
  scripts that direct the core parts of my project
  they rely on functions and scripts within 'functions'

### masterscripts

all scripts that drive the core transcriptional profiling using co-expression and differential expression

| file | function |
| ------ | ----------- |
| coexpression_method_comparison_master.R   | Compares three co-expression methods (Self Organising Maps, Weighted Gene Co-expression Analysis and k-means) |
| differential_expression_master.R | three different differential expression methodologies (DESeq2, limmaR and sleuth)|
| kallisto.sh    | raw sequence read alignment and mapping using fasta paired end sequencefiles  |
| self_organising_map_.R    | self-organising map to transcriptionally profile cell subsets  |
| wgcna_master.R    | self-organising map to transcriptionally profile cell subsets  |

### functions

contains all functions and scripts that master scripts direct to produce results





